### PR TITLE
Classification (new package)

### PR DESCRIPTION
**Pull request by sgiraudot**
## Summary of Changes

This is a new package. Not yet approved [full feature](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Classification).

## Release Management

* Affected package(s): Classification (new package), Polyhedron demo
* Feature: https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Classification

### CODE REVIEW

**Comment by lrineau** [comment]
The test results are almost all red:
https://cgal.geometryfactory.com/CGAL/testsuite/results-4.12-Ic-27.shtml#Classification

**Comment by lrineau** [comment]
... and the Travis test reported:
  - a pb with the `CHECK` line,
  - and with the Polyhedron_Demo line.
https://travis-ci.org/CGAL/cgal/builds/269101674?utm_source=github_status&utm_medium=notification

**Comment by sgiraudot** [comment]
Most errors come from a file I forgot to push, this should be fixed.

I need help for the license part: I generated the `license/Classification.h` file using `Installation/include/CGAL/license/generate_files.cmake` and included it in all headers of the new package. Is there anything else I need to do?

**Comment by lrineau** [comment]
And the classification plugin should be added to `demo/Polyhedron/cgal_test_with_cgal`, so that has its own line in the summary of `success`/`ERROR` at the beginning of each test result.

**Comment by lrineau** [comment]
As for the license files, I think you did right. @afabri, can you confirm?

**Comment by lrineau** [comment]
As for licenses, the file `Classification/include/CGAL/Classification/Random_forest_classifier.h` does not have a proper license notice. That is why the Travis line `CHECK` fails.

**Comment by sgiraudot** [comment]
I had put it after some other includes, don't know why… should be fixed now.

**Comment by lrineau** [comment]
What's missing in that file is the copyright/license notices. The license-checking header is probably right.
```
// Copyright (c) 2017 GeometryFactory Sarl (France).
// All rights reserved.
//
// This file is part of CGAL (www.cgal.org).
// You can redistribute it and/or modify it under the terms of the GNU
// General Public License as published by the Free Software Foundation,
// either version 3 of the License, or (at your option) any later version.
//
// Licensees holding a valid commercial license may use this file in
// accordance with the commercial license agreement provided with the software.
//
// This file is provided AS IS with NO WARRANTY OF ANY KIND, INCLUDING THE
// WARRANTY OF DESIGN, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
//
```

**Comment by sgiraudot** [comment]
Oh, right, thanks. I fixed it too.

**Comment by lrineau** [comment]
There is an error in the CHECK line of Travis: https://travis-ci.org/CGAL/cgal/jobs/269506998#L2487

**Comment by sgiraudot** [comment]
Still an error in Travis, I'm investigating.

**Comment by lrineau** [comment]
In addition to the compilation errors, there are warnings, on recent MSVC (that manages to compile the code):
https://cgal.geometryfactory.com/CGAL/testsuite/results-4.12-Ic-29.shtml#Classification

File: Polyhedron/demo/Polyhedron/Plugins/Classification/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,19 @@
+include( polyhedron_demo_macros )
+
+if(EIGEN3_FOUND)
+
+  qt5_wrap_ui( classificationUI_FILES Classification_widget.ui)
+  polyhedron_demo_plugin(classification_plugin Classification_plugin Point_set_item_classification ${classificationUI_FILES})
+  find_package(OpenCV)
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Plugins/Classification/CMakeLists.txt]: CMake warnings: please add the `QUIET` here as well

**Comment by sgiraudot** [comment]
Okay, I finally understood the Travis error… turns out the `xml_writer_settings` (that is used in Boost examples but has basically 0 documentation) takes a `char` type in old versions of Boost (like in Travis) and `string` type in recent versions (like in my config). I had to check the sources of Boost to understand this…

I'll see how I fix it.

**Comment by sloriot** [comment]
The doc for the license checking mechanism is [here](https://github.com/cgal/cgal/wiki/License)

**Comment by lrineau** [comment]
The last testsuite was not so successful: https://cgal.geometryfactory.com/CGAL/testsuite/results-4.12-Ic-31.shtml#Classification

**Comment by sgiraudot** [comment]
Still some warnings… I'll fix them.

**Comment by lrineau** [comment]
You can try to compile with `-Wconversion` but it might be painful, because the flag `-Wconversion` (at least with gcc) is more picky than the warnings of MSVC.

I did the job for the whole Mesh_3 testsuite, to make sure I can use `-Wconversion` and detect real conversion problems.

**Comment by sgiraudot** [comment]
There was only one warning remaining in an example, so it should be good now (I amended the latest commit as it was the same kind of corrections).

**Comment by lrineau** [comment]
There are still warnings, in examples, tests, and in the classification plugin of the 3D demo:
- https://cgal.geometryfactory.com/CGAL/testsuite/results-4.12-Ic-36.shtml#Classification
- https://cgal.geometryfactory.com/CGAL/testsuite/results-4.12-Ic-36.shtml#Polyhedron_Demo

File: Installation/changes.html [hunk 2]
```diff
@@ -136,6 +136,14 @@ <h2 id="release4.12">Release 4.12 </h2>
 
 <!-- Installation (and general changes) -->
 <!-- New packages -->
+<h3>Classification</h3>
```
Comment (lrineau) [inline Installation/changes.html]: It should be `Classification (new package)`.

File: Installation/changes.html [hunk 3]
```diff
@@ -136,6 +136,14 @@ <h2 id="release4.12">Release 4.12 </h2>
...
+<ul>
+  <li>This package offers an algorithm that classifies a data set into
+  a user-defined set of labels (such as ground, vegetation, buildings,
+  etc.). A flexible API is provided so that users can classify any
+  type of data, compute their own local features on the input data set
+  and define their own labels.</li>
```
Comment (lrineau) [inline Installation/changes.html]: Add a `,` before the last `and`.

**Comment by lrineau** [comment]
@sloriot The weight of this PR is 2.5MB. Can we do something?

**Comment by sgiraudot** [comment]
I've been uploading different versions of the figures _a lot_, I don't know how to do it but I know @sloriot has a trick to fix this…

**Comment by sloriot** [comment]
Down to 1.5MB now

**Comment by lrineau** [comment]
There are error in the Classification results, now:
https://cgal.geometryfactory.com/CGAL/testsuite/results-4.12-Ic-43.shtml#Classification_Examples

**Comment by sloriot** [comment]
@sgiraudot if it is an issue with the binary vs ASCII ply we should be able to easily revert that change if needed

**Comment by sgiraudot** [comment]
I'm investigating.

**Comment by sgiraudot** [comment]
It was indeed a binary problem, Windows expects an implicit use of `std::ios::binary`. It's fixed now.

**Comment by lrineau** [comment]
This PR was correctly tested in CGAL-4.12-Ic-45, and it seems nothing is blocking its integration into `master`, now.

**Pull request merged**